content/advent-2018/go-in-the-browser.md

```diff
@@ -0,0 +1,175 @@
+
+So far so good. But this was a simple hello world program. A real world webapp needs to interact with the DOM. We need to respond to button click events, take input data from text boxes, and send data back to the DOM. Now we will build a minimal image editor in the browser which will use all of these capabilities.
+
+## DOM API
+
+But first, to interact with the browser from Go code, we need a DOM API. We have the `syscall/js` library to help us out with that. It is a very basic, but nevertheless powerful form of a DOM API, on top of which we can build our app. Let us very quickly see some of its capabilities before we move over to making our app.
+
+**Callbacks**
+
+To respond to DOM events, we declare callbacks and hook them up with events like this -
+
+```
+import “syscall/js”
+
+// Declare callback
+cb := js.NewEventCallback(js.PreventDefault, func(ev js.Value) {
+	// handle event
+})
+
+
+// Hook it up with a DOM event
+js.Global().Get("document").
+	Call("getElementById", "myBtn").
+	Call("addEventListener", "click", cb)
+
+
+// Call cb.Release() on your way out.
+```
+
+**Updating the DOM**
+
+To update the DOM from inside Go, we can do-
+
+```
+import “syscall/js”
+
+js.Global().Get("document").
+		Call("getElementById", "myTextBox").
+		Set("value", "hello wasm")
+```
+
+You can even call JS functions and manipulate native native JS Objects like `FileReader` or `Canvas`. Feel free to check out the [syscall/js](https://golang.org/pkg/syscall/js/) documentation for further details.
+
+Ok, now on with building our app !
+
+## A proper webapp
+
+We will build a small app which will take an input image, then perform some manipulations on the image like brightness, contrast, hue, saturation, and finally send the output image back to the browser. There will be sliders for each of these effects, which the user can change and see the target image change in real time.
+
+First, we need to get the input image from the browser to our Go code, so that we can work on it. To efficiently do this, we need to resort to some `unsafe` tricks, the details of which I will skip here. Once we have the image, it is fully in our control and we are free to do whatever with it. Below is a brief snippet from the image loader callback -
+
+```go
+s.onImgLoadCb = js.NewCallback(func(args []js.Value) {
+	reader := bytes.NewReader(s.inBuf)
+	var err error
+	s.sourceImg, _, err = image.Decode(reader)
+	if err != nil {
+		s.log(err.Error())
+		return
+	}
+	// Now the sourceImg is an image.Image with which we are free to do anything !
+})
+```
+
+Then we take user values from any of the effect sliders, and manipulate the image. We use the awesome [bild](https://github.com/anthonynsimon/bild) library for that. Here is a small snippet of the contrast callback -
+
+```go
+s.contrastCb = js.NewEventCallback(js.PreventDefault, func(ev js.Value) {
+	delta := ev.Get("target").Get("valueAsNumber").Float()
+	res := adjust.Contrast(s.sourceImg, delta)
```

Comment:
How about you add `import github.com/anthonynsimon/bild/adjust` to the snippet? It'll make easier to follow the code as there are lots of stuff happening here.

Reply:
Yep !